samza-core/src/main/java/org/apache/samza/table/caching/CachingTable.java

```diff
@@ -0,0 +1,152 @@
+   * we synchronize the whole table instead.
+   */
+  private final ConcurrentHashMap<K, V> locker;
+
+  public CachingTable(ReadableTable<K, V> table, ReadWriteTable<K, V> cache, boolean isWriteAround) {
+    this.rdTable = table;
+    this.rwTable = table instanceof ReadWriteTable ? (ReadWriteTable) table : null;
+    this.cache = cache;
+    this.isWriteAround = isWriteAround;
+    this.locker = new ConcurrentHashMap();
+  }
+
+  @Override
+  public V get(K key) {
+    V value = cache.get(key);
+    if (value != null) {
+      return value;
+    }
+
+    // Atomically populates the value for a cache miss
+    locker.computeIfAbsent(key, (k) -> {
+        // Recheck in case it is already populated by another thread
+        if (cache.get(key) == null) {
+          // Due to the lack of contains() API in ReadableTable, there is
+          // no way to tell whether a null return by cache.get(key) means
+          // cache miss or the value is actually null. As such, we cannot
+          // support negative cache semantics.
+          cache.put(key, rdTable.get(key));
+        }
+        // Discard the value so it is not entered into the locker map
+        return null;
+      });
+
+    return cache.get(key);
+  }
+
+  @Override
+  public Map<K, V> getAll(List<K> keys) {
+    Map<K, V> retMap = new HashMap<>();
+    keys.stream().forEach(k -> retMap.put(k, get(k)));
+    return retMap;
+  }
+
+  @Override
+  public void close() {
+    this.cache.close();
+    this.rdTable.close();
+  }
+
+  @Override
+  public void put(K key, V value) {
+    Preconditions.checkNotNull(rwTable, "Cannot write to a read-only table: " + rdTable);
+    if (isWriteAround) {
+      this.rwTable.put(key, value);
+    } else {
+      locker.computeIfAbsent(key, (k) -> {
+          this.rwTable.put(key, value);
+          this.cache.put(key, value);
+          // Discard the value so it is not entered into the locker map
+          return null;
+        });
+    }
+  }
+
+  @Override
+  public synchronized void putAll(List<Entry<K, V>> entries) {
+    Preconditions.checkNotNull(rwTable, "Cannot write to a read-only table: " + rdTable);
+    entries.forEach(e -> {
+        this.rwTable.put(e.getKey(), e.getValue());
+        this.cache.put(e.getKey(), e.getValue());
+      });
+  }
+
+  @Override
+  public void delete(K key) {
+    Preconditions.checkNotNull(rwTable, "Cannot delete from a read-only table: " + rdTable);
+    locker.computeIfAbsent(key, (k) -> {
+        cache.delete(key);
+        rwTable.delete(key);
+        return null;
+      });
+  }
+
+  @Override
+  public synchronized void deleteAll(List<K> keys) {
+    Preconditions.checkNotNull(rwTable, "Cannot delete from a read-only table: " + rdTable);
+    keys.stream().forEach(k -> {
+        cache.delete(k);
+        rwTable.delete(k);
+      });
+  }
+
+  @Override
+  public synchronized void flush() {
+    Preconditions.checkNotNull(rwTable, "Cannot flush a read-only table: " + rdTable);
```

Comment:
I think we should throw because this indicates a configuration problem, ie. app is expecting a RW table but only configured a RO table. Does this make sense?